packages/kit/src/runtime/client/router.js

```diff
@@ -190,9 +205,17 @@ export class Router {
 			// Removing the hash does a full page navigation in the browser, so make sure a hash is present
 			const [base, hash] = url.href.split('#');
 			if (hash !== undefined && base === location.href.split('#')[0]) {
-				// Call `pushState` to add url to history so going back works.
-				// Also make a delay, otherwise the browser default behaviour would not kick in
-				setTimeout(() => history.pushState({}, '', url.href));
+				this.#save_scroll_position();
+
+				// Call `replaceState` in timeout because this event will add pushState automatically and we need to store our own history state
+				// Timeout is being used to delay replaceState till next tick
+				setTimeout(() => {
+					history.replaceState(
+						{ ...history.state, 'sveltekit:index': ++this.current_history_index },
```

Comment:
why are we storing `sveltekit:index` in the history state here and in the session elsewhere?

Reply:
why not just store the scroll state in the history directly?